### PR TITLE
ReservedPackage + admin action.

### LINKED ISSUE #8082: Admin action for reserving package names
Currently we need a new deployment for reserving a name.

Would be great to have a dynamic system for this. A datastore entry and an action

### PR DESCRIPTION
**Pull request by isoos**
- Fixes #8082

### CODE REVIEW

File: app/lib/admin/actions/package_reserve.dart [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  description: '''
+Reserves a package name that can be claimed by @google.com accounts or
+the allowed list of email addresses.
+
+The action can be re-run with the same package name. In such cases the provided
+email list will be added to the existing ReservedPackage entity.
+''',
+  options: {
+    'package': 'The package name to be reserved.',
+    'emails': 'The list of email addresses, separated by comma.'
```
Comment (sigurdm) [inline app/lib/admin/actions/package_reserve.dart:23]: Should we add an option to delete the entity? Maybe it is not worth it...

**Review by sigurdm — APPROVED** [review]

File: app/lib/admin/actions/package_reserve.dart [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    if (mp != null) {
+      throw InvalidInputException('ModeratedPackage `$package` exists.');
+    }
+
+    final entry = await withRetryTransaction(dbService, (tx) async {
+      final existing = await tx.lookupOrNull<ReservedPackage>(
+          dbService.emptyKey.append(ReservedPackage, id: package));
+      final entry = existing ?? ReservedPackage.init(package);
+      if (emails != null) {
+        entry.emails = <String>{...entry.emails, ...emails}.toList();
```
Comment (jonasfj) [inline app/lib/admin/actions/package_reserve.dart:47]:
```suggestion
        entry.emails = <String>{...emails}.toList();
```
I think it's safer to overwrite than append.

File: app/lib/admin/actions/package_reserve.dart (hunk 1 above)
Comment (jonasfj) [reply app/lib/admin/actions/package_reserve.dart:23]: We could just add:
 * `package-reserve-list`
 * `package-reserve-delete`

Life is easier, if we split them into multiple actions. We can still put them all in the same file if want that.

We could even consider naming them: 
 `package-reservation-create`
 `package-reservation-list`
 `package-reservation-delete`

naming is a nit, but it's probably more intuitive to have separate options. And it's probably easier to read the code too.

Comment (isoos) [reply app/lib/admin/actions/package_reserve.dart:23]: Like a `--delete` flag?

File: app/lib/package/backend.dart [hunk 3]
```diff
@@ -1014,10 +1022,19 @@ class PackageBackend {
...
-    final isReservedName = matchesReservedPackageName(name);
-    final isExempted = isGoogleComUser && isReservedName;
+    final reservedPackage = await lookupReservedPackage(name);
+    final reservedEmails = reservedPackage?.emails ?? const <String>[];
+
+    bool isAllowedUser = false;
+    if (agent is AuthenticatedUser) {
+      final email = agent.user.email;
+      isAllowedUser = email != null &&
+          (email.endsWith('@google.com') || reservedEmails.contains(email));
```
Comment (jonasfj) [inline app/lib/package/backend.dart:1032]: I suggest that now we can allow list specific emails, we don't do this general thing for `@google.com` emails.

It's much better to be specific, this is generally rare.

File: app/lib/admin/actions/package_reserve.dart (hunk 1 above)
Comment (sigurdm) [reply app/lib/admin/actions/package_reserve.dart:23]: Yeah - probably `--delete=true` as we didn't implement real flags for actions,

Comment (jonasfj) [reply app/lib/admin/actions/package_reserve.dart:23]: a separate command is MUCH easier to do.

Comment (jonasfj) [reply app/lib/admin/actions/package_reserve.dart:23]: it's really just something like:
```dart
final package = options['package'];
InvalidInputException.check(
  package != null && package.isNotEmpty,
  '`package` must be given',
);

await withRetryTransaction(dbService, (tx) async {
  final existing = await tx.lookupOrNull<ReservedPackage>(
      dbService.emptyKey.append(ReservedPackage, id: package));
      
  if (exiting == null) {
    return {'message': 'ReservedPackage for "$package" does not exist!'};
  }
  
  tx.delete(existing.key);
});

return {'message': 'ReservedPackage for "$package" deleted.'};
```

A `--delete=true` flag requires a lot more explanation in the documentation, it's not very intuitive.

**Comment by sigurdm** [comment]
Yeah - multiple commands is probably better.

**Comment by isoos** [comment]
Updated, PTAL.

**Comment by isoos** [comment]
I will add list/delete in a follow-up PR.

**Review by jonasfj — APPROVED** [review]

**Review by sigurdm — APPROVED** [review]

**Pull request merged**